src/uart.rs

```diff
@@ -665,6 +792,217 @@ impl<'d> core::fmt::Write for UartTxDriver<'d> {
+            stop_bits.into()
+        )
+    }
+
+    fn data_bits(&self) -> Result<config::DataBits, EspError> {
+        let mut data_bits: uart_word_length_t = 0;
+        esp_result!(
+            unsafe { uart_get_word_length(self.port(), &mut data_bits) },
+            data_bits.into()
+        )
+    }
+
+    fn parity(&self) -> Result<config::Parity, EspError> {
+        let mut parity: uart_parity_t = 0;
+        esp_result!(
+            unsafe { uart_get_parity(self.port(), &mut parity) },
+            parity.into()
+        )
+    }
+
+    fn baudrate(&self) -> Result<Hertz, EspError> {
+        let mut baudrate: u32 = 0;
+        esp_result!(
+            unsafe { uart_get_baudrate(self.port(), &mut baudrate) },
+            baudrate.into()
+        )
+    }
+}
+
+impl<'d> Port<'d, Owned> {
+    fn borrow(&mut self) -> Port<'d, Borrowed> {
+        Port {
+            port: Borrowed(self.port() as _),
+            _p: PhantomData,
+        }
+    }
+
+    fn change_stop_bits(&mut self, stop_bits: config::StopBits) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_stop_bits(self.port(), stop_bits.into()) })
+    }
+
+    fn change_data_bits(&mut self, data_bits: config::DataBits) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_word_length(self.port(), data_bits.into()) })
+    }
+
+    fn change_parity(&mut self, parity: config::Parity) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_parity(self.port(), parity.into()) })
+    }
+
+    fn change_baudrate<T: Into<Hertz> + Copy>(&mut self, baudrate: T) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_baudrate(self.port(), baudrate.into().into()) })
+    }
+}
+
+mod sealed {
+    pub trait AsPort {
+        fn port(&self) -> esp_idf_sys::uart_port_t;
+    }
+}
+
+pub struct Owned(u8);
+
+impl sealed::AsPort for Owned {
+    fn port(&self) -> uart_port_t {
+        self.0 as _
+    }
+}
+
+impl Drop for Owned {
+    fn drop(&mut self) {
+        esp!(unsafe { uart_driver_delete(self.0 as _) }).unwrap()
+    }
+}
+
+pub struct Borrowed(u8);
+
+impl sealed::AsPort for Borrowed {
+    fn port(&self) -> uart_port_t {
+        self.0 as _
+    }
+}
+
+#[cfg(feature = "alloc")]
+mod shared {
+    use super::*;
+
+    extern crate alloc;
+    use alloc::sync::Arc;
+
+    pub struct Shared(Arc<Owned>);
+
+    impl<'d> UartDriver<'d> {
+        /// Split the serial driver in separate TX and RX drivers.
+        ///
+        /// Unlike [`split`], the halves are owned, at the cost of an allocation.
+        pub fn into_split(self) -> (UartTxDriver<'d, Shared>, UartRxDriver<'d, Shared>) {
+            let shared = self.0.into_shared();
+            (UartTxDriver(shared.clone()), UartRxDriver(shared))
+        }
+
+        /// Merge the shared TX and RX halves into an owned [`UartDriver`].
+        ///
+        /// The parts must be produced from the same call to [`into_split`].
+        pub fn merge_split(
```

Comment:
I think if `'d` was an invariant lifetime, `ptr_eq` wouldn't be necessary and no need to return a `Result`, just unwrap. But I'm not sure if it works with `'static`.

Reply:
I don't think we really need `merge_split`. I'm not doing this anywhere, and it somehow does not fit the new philosophy of 
* Moving (i.e. `into_split` or passing peripherals by moving/owning) is irreversible
* If you want to reverse - pass by `&mut ref` (i.e. `split` or passing peripherals by `&mut ref`